### PR TITLE
Cache preprocessed speech features instead of recomputing on-the-fly

### PR DESCRIPTION
**Pull request by mvankeirsbilck**
For networks where network evaluation is fast, on-the-fly preprocessing can take a significant fraction of runtime.  Storing these preprocessed features to disk avoids having to recompute them all the time.
Especially for MFCC features preprocessing takes a long time.

For the provided example network, a 2-layer RNN (512 units/layer), evaluation and training are >5x faster with caching using 13 MFCC features than with on-the-fly preprocessing.
Without caching, CPU usage is very high (>90%) and GPU usage very low (<10%).
With caching, the CPU is no longer the bottleneck.

For larger networks (eg. ds2_small_1gpu), there is little difference as the GPU is the bottleneck there.

3 parameters configure the caching:
- cache_features: turn caching on or off
- cache_format: format to save the preprocessed data in (default HDF5)
- cache_regenerate: force recomputation of preprocessed features (only if cache_features==True)

The preprocessed features are stored in the dataset location, next to the 'wav' folder.
The preprocessed-folder is named using the config parameters specified for preprocessing.
There is no seperate preprocessing step required; the features are generated and stored during the first run (first epoch).

Probably higher speedups are possible using more compact formats, as there is still one file per WAV.
However this is a simple and flexible way of getting good speedups for small/fast networks.

### CODE REVIEW

**Comment by GabrielLin** [comment]
It is very useful. I hope this PR this be merged.

File: example_configs/speech2text/lstm_small_1gpu.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
+# pylint: skip-file
+import os
+<<<<<<< HEAD
```
Comment (vsl9) [inline example_configs/speech2text/lstm_small_1gpu.py]: Can you please remove lines 3-6?

**Comment by vsl9** [comment]
Can you please sign-off your 3rd commit: 9cf200f135e85c61250075e93702c2cab3f33058?

**Comment by mvankeirsbilck** [comment]
Hi,
sorry for the delay, I was working on some other issues and lost track of this.
I squashed the last commits, and signed them. I think it should be good now?

**Pull request merged**